File: conformance/service-1/components/provider.ts
import {
  OperationError,
  ServiceError,
  serveService,
  type ServiceDefinition,
} from "../../../packages/flow-sdk/src/index.ts";

const service: ServiceDefinition = {
  exports: process.env.FLOWMD_TEST_EMPTY_SERVICE === "1" ? {} : {
    sessions: async (context) => {
      if (context.method === "echo") return context.input;
      if (context.method === "missing") throw new ServiceError("not-found", context.input);
      if (context.method === "dependency") {
        return await context.callEffect({
          operationId: "storage:1",
          slot: "storage",
          method: "read",
          input: context.input,
        });
      }
      if (context.method === "detached") {
        void context.callEffect({
          operationId: "detached:1",
          slot: "storage",
          method: "read",
          input: context.input,
        });
        return "must-not-succeed";
      }
      if (context.method === "fanout") {
        return await Promise.all(Array.from({ length: 65 }, async (_, index) => {
          try {
            await context.callEffect({
              operationId: `fanout:${index}`,
              slot: "storage",
              method: "read",
              input: index,
            });
            return "ok";
          } catch (error) {
            return error instanceof OperationError ? error.code : "unexpected";
          }
        }));
      }
      if (context.method === "slow") {
        await new Promise<void>((resolve) => {
          context.signal.addEventListener("abort", () => resolve(), { once: true });
        });
        return "caught-cancellation";
      }
      return null;
    },
  },
  async mount(context) {
    if (context.settings.initialize === true) {
      await context.callEffect({
        operationId: "initialize:1",
        slot: "storage",
        method: "open",
        input: null,
      });
    }
    await context.ready();
    if (context.settings.detachedMount === true) {
      void context.callEffect({
        operationId: "mount-detached:1",
        slot: "storage",
        method: "read",
        input: null,
      });
      return;
    }
    await context.cancelled;
  },
};

await serveService(service);

File: conformance/service-1/components/provider.py
from __future__ import annotations

import asyncio
import os

from flowmd_sdk.service import OperationError, ServiceDefinition, ServiceError, serve_service


async def sessions(context):
    if context.method == "echo":
        return context.input
    if context.method == "missing":
        raise ServiceError("not-found", context.input)
    if context.method == "dependency":
        return await context.call_effect(
            operation_id="storage:1",
            slot="storage",
            method="read",
            input=context.input,
        )
    if context.method == "detached":
        task = asyncio.create_task(context.call_effect(
            operation_id="detached:1",
            slot="storage",
            method="read",
            input=context.input,
        ))
        task.add_done_callback(lambda completed: completed.exception())
        await asyncio.sleep(0)
        return "must-not-succeed"
    if context.method == "fanout":
        settled = await asyncio.gather(*(
            context.call_effect(
                operation_id=f"fanout:{index}",
                slot="storage",
                method="read",
                input=index,
            )
            for index in range(65)
        ), return_exceptions=True)
        return [
            item.code if isinstance(item, OperationError) else "ok"
            for item in settled
        ]
    if context.method == "slow":
        try:
            await asyncio.Event().wait()
        except asyncio.CancelledError:
            return "caught-cancellation"
    return None


async def mount(context):
    if context.settings.get("initialize") is True:
        await context.call_effect(
            operation_id="initialize:1",
            slot="storage",
            method="open",
            input=None,
        )
    await context.ready()
    if context.settings.get("detachedMount") is True:
        task = asyncio.create_task(context.call_effect(
            operation_id="mount-detached:1",
            slot="storage",
            method="read",
            input=None,
        ))
        task.add_done_callback(lambda completed: completed.exception())
        await asyncio.sleep(0)
        return
    await context.cancelled()


serve_service(ServiceDefinition(
    exports={} if os.environ.get("FLOWMD_TEST_EMPTY_SERVICE") == "1" else {"sessions": sessions},
    mount=mount,
))
